### PR TITLE
PanelOptions: Refactoring applying panel and field options out of PanelModel and add property clean up for properties not in field config registry 

### LINKED ISSUE #29895: PanelEdit: Clean field config properties when switching panel plugin
We currently clean all top-level properties and the panel options when switching visualization.  We also remember these options so when switching back the are restored. 

We hover preserve all field config options. The main reason is that we want to preserve the standard field options. 

TODO
* [ ] Clean/remove any custom field config options from defaults and overrides. And remember them so we can restore them if we switch back to the same visualisation
* [ ] Clean/remove any standard field options that are not used by the new panel, remember them in case switching to a new panel (or back) that does. 

Relevant area where this logic is: 
https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelModel.ts#L373

### PR DESCRIPTION
**Pull request by torkelo**
PanelModel is getting too big, this is moving the options & field config defaults & option merging (and validation) out of PanelModel to simplify the code & tests and reduce the size of PanelModel. 

* [x] Move apply field and options out 
* [x] Move tests 
* [x] Write some new tests
* [x] Handle cleaning custom properties that are not part of the registry 
* [x] Handle cleaning out overrides and overrides with no properties
* [x] How to restore options when switching back and forth between visualizations 
* [x] Clear all custom options & custom overrides before switch
* [x] Remember and restore all custom defaults and overrides 

Closes #29895

### CODE REVIEW

File: packages/grafana-data/src/panel/PanelPlugin.ts [hunk 1]
```diff
@@ -137,6 +139,10 @@ export class PanelPlugin<TOptions = any, TFieldConfigOptions extends object = an
     configDefaults.custom = {} as TFieldConfigOptions;
 
     for (const option of this.fieldConfigRegistry.list()) {
+      if (option.defaultValue === null || option.defaultValue === undefined) {
+        continue;
+      }
+
       set(configDefaults, option.id, option.defaultValue);
     }
```
Comment (torkelo) [inline packages/grafana-data/src/panel/PanelPlugin.ts:147]: We where setting props with a default value of undefined, maybe we should set props with a default value of null as there is a distinction there

Comment (torkelo) [reply packages/grafana-data/src/panel/PanelPlugin.ts]: changed this to only ignore props with undefined as default

File: packages/grafana-data/src/types/panel.ts [hunk 2]
```diff
@@ -131,7 +131,8 @@ export type PanelMigrationHandler<TOptions = any> = (panel: PanelModel<TOptions>
 export type PanelTypeChangedHandler<TOptions = any> = (
   panel: PanelModel<TOptions>,
   prevPluginId: string,
-  prevOptions: any
+  prevOptions: Record<string, any>,
+  prevFieldConfig: FieldConfigSource
```
Comment (torkelo) [inline packages/grafana-data/src/types/panel.ts:135]: adding the previous panel's field config here to the PanelTypeChangedHandler @ryantxu

File: public/app/features/dashboard/state/getPanelOptionsWithDefaults.test.ts [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+                ],
+              },
+              "unit": "bytes",
+            },
+            "overrides": Array [],
+          },
+          "options": Object {
+            "name": "hello",
+            "number": 20,
+            "showThresholds": false,
```
Comment (dprokop) [inline public/app/features/dashboard/state/getPanelOptionsWithDefaults.test.ts:162]: Ok, so the options that are currently set are preserved and the rest is set to defaults?

File: public/app/features/dashboard/state/getPanelOptionsWithDefaults.test.ts [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+              preferThresholdsMode: true,
+            },
+          },
+        },
+      });
+      expect(result.fieldConfig.defaults.color!.mode).toBe(FieldColorModeId.Thresholds);
+    });
+  });
+
+  describe('when applying defaults clean properties that no longer part of the registry', () => {
```
Comment (dprokop) [inline public/app/features/dashboard/state/getPanelOptionsWithDefaults.test.ts:231]:
```suggestion
  describe('when applying defaults clean properties that are no longer part of the registry', () => {
```

**Review by dprokop — APPROVED** [review]
This works as expected and just wow, I don't know how you can shoot with such PRs with this speed :D The color mode settings changes were slightly tricky to understand, but other than that beautiful PR.

File: public/app/features/dashboard/state/getPanelOptionsWithDefaults.test.ts (hunk 3 above)
Comment (torkelo) [reply public/app/features/dashboard/state/getPanelOptionsWithDefaults.test.ts:162]: yes

**Pull request merged**